javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/collector/ReferenceCollector.java

```diff
@@ -201,6 +203,73 @@ private void collectHelperClasses(
     return references;
   }
 
+  public void prune() {
+    Set<Reference> helperClassesWithLibrarySuperType = getHelperClassesWithLibrarySuperType();
+
+    Set<Reference> needToKeepFieldsAndMethods = new HashSet<>();
+    for (Reference reference : helperClassesWithLibrarySuperType) {
+      addSuperClasses(reference.getClassName(), needToKeepFieldsAndMethods);
+    }
+
+    for (Iterator<Map.Entry<String, Reference>> i = references.entrySet().iterator();
+        i.hasNext(); ) {
+      Reference reference = i.next().getValue();
+      if (instrumentationClassPredicate.isProvidedByLibrary(reference.getClassName())) {
+        // these are the references to library classes which need to be checked at runtime
+        continue;
+      }
+      if (needToKeepFieldsAndMethods.contains(reference)) {
+        // these need to be kept in order to check abstract methods are implemented and declared
+        // super class fields are present
+        // TODO (trask) if this is provided by javaagent then can remove methods since the whole
+        //  class will be resolved at runtime
+        //  or another option is to resolve from helper classes first, see
+        //  HelperReferenceWrapper.Factory.create(String)
```

Comment:
Helper classes are not loaded when the muzzle matcher executes; they're just resources somewhere in the agent CL.

Reply:
I thought this comment made sense at the time... 😂